### PR TITLE
NAS-113493 / 22.02-RC.2 / NAS-113493: Add unit tests for Services

### PR DESCRIPTION
**Pull request by denysbutenko**

### CODE REVIEW

File: src/app/pages/services/services.component.html [hunk 1]
```diff
@@ -2,63 +2,61 @@
   <mat-card-content>
     <table ix-table [dataSource]="dataSource">
       <ng-container matColumnDef="name">
-        <th *matHeaderCellDef>{{ 'Name' | translate }}</th>
-        <td *matCellDef="let service; dataSource: dataSource">{{ serviceNames.get(service.service) }}</td>
+        <th mat-header-cell *matHeaderCellDef>{{ 'Name' | translate }}</th>
```
Comment (undsoft) [inline src/app/pages/services/services.component.html:5]: I'm assuming we need those for tests. That's too bad :)

File: src/app/pages/services/services.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+  beforeEach(() => {
+    spectator = createComponent();
+    loader = TestbedHarnessEnvironment.loader(spectator.fixture);
+    ws = spectator.inject(WebSocketService);
+    spectator.fixture.detectChanges();
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
```
Comment (undsoft) [inline src/app/pages/services/services.component.spec.ts:71]: Why is this necessary?

File: src/app/pages/services/services.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    const table = await loader.getHarness(IxTableHarness);
+    const headerRow = await table.getHeaderRow();
+
+    expect(headerRow).toMatchObject({
+      name: 'Name',
+      state: 'Running',
+      enable: 'Start Automatically',
+      actions: '',
+    });
+  });
```
Comment (undsoft) [inline src/app/pages/services/services.component.spec.ts:84]: You are kinda testing this already in next test case.

File: src/app/pages/services/services.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    const cells = await table.getCells(true);
+    const expectedData = [...serviceNames.keys()]
+      .filter((service) => ![ServiceName.Gluster, ServiceName.Afp].includes(service))
+      .map((service) => [serviceNames.get(service), '', '', 'edit']);
+    const expectedRows = [['Name', 'Running', 'Start Automatically', ''], ...expectedData];
+
+    expect(cells).toEqual(expectedRows);
+  });
+
+  it('should redirect to configure service page', async () => {
```
Comment (undsoft) [inline src/app/pages/services/services.component.spec.ts:97]: ...when edit button is pressed.

File: src/app/pages/services/services.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    const editButton = await loader.getHarness(MatButtonHarness.with({ text: 'edit' }));
+    await editButton.click();
+
+    expect(spectator.inject(Router).navigate).toHaveBeenCalledWith(['/services', 'dynamicdns']);
```
Comment (undsoft) [inline src/app/pages/services/services.component.spec.ts:101]: Test will be a bit fragile if we add another service to the top of enum.

File: src/app/pages/services/services.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  it('should check service state changes', async () => {
+    const slideToggle = await loader.getHarness(MatSlideToggleHarness);
+    await slideToggle.toggle();
+
+    expect(await slideToggle.isChecked()).toBeTruthy();
+    expect(ws.call).toHaveBeenCalledWith('service.start', ['dynamicdns']);
+  });
+
+  it('should check service autostart changes', async () => {
```
Comment (undsoft) [inline src/app/pages/services/services.component.spec.ts:112]: I'd prefer something like 'should change service autostart state when checkbox is ticked'.
Same for one above.

**Review by undsoft — APPROVED** [review]
Looks good.

File: src/app/pages/services/services.component.html (hunk 1 above)
Comment (denysbutenko) [reply src/app/pages/services/services.component.html:5]: Yeah, exactly. Otherwise, the table harness couldn't get values
